// @minka/ledger-sdk/src/lib/common/services/secret-reveal-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretRevealBuilder = void 0;
const ledger_sdk_error_1 = require("../types/ledger-sdk-error");
const base_record_builder_1 = require("./base-record-builder");
/**
 * Implements a builder pattern that allows a more
 * convenient way to work with ledger records. This
 * class has functions that simplify construction,
 * hashing, signing, testing and sending records to
 * remote ledger instances.
 *
 * This builder is intended for sending secret reveal
 * queries to ledger, and will attempt to send it on
 * the API with the content that was prepared in the
 * builder.
 */
class SecretRevealBuilder extends base_record_builder_1.BaseRecordBuilder {
    /**
     * A handler for secret reveal on the remote API.
     */
    revealSecret;
    /**
     * Creates new instance of secret reveal builder.
     *
     * @param revealSecret function which handles secret reveal on ledger api
     */
    constructor(revealSecret) {
        super();
        this.revealSecret = revealSecret;
    }
    /**
     * Sends the current secret query payload to the remote ledger.
     *
     * @param requestOptions Optional request options to use for this request
     * @returns api response with the decrypted secrets
     * @throws if any method in builder chain fails
     */
    async send(requestOptions) {
        try {
            await this.finishAsyncOperations();
            const finalOptions = {
                authParams: requestOptions?.authParams ?? this.authParams,
                headers: requestOptions?.headers,
            };
            return this.revealSecret(this.record, finalOptions);
        }
        catch (error) {
            throw new ledger_sdk_error_1.LedgerSdkError(error);
        }
    }
}
exports.SecretRevealBuilder = SecretRevealBuilder;
