// @minka/ledger-sdk/types/src/lib/records/ledger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LedgerConfigKey = void 0;
var LedgerConfigKey;
(function (LedgerConfigKey) {
    LedgerConfigKey["IntentExpiryThresholdMinutes"] = "intent.expiryThresholdMinutes";
    LedgerConfigKey["IntentExpiryThresholdMinutesDeprecated"] = "intentExpiryThresholdMinutes";
    LedgerConfigKey["IntentCreditorInitiatedTransfersEnabled"] = "intent.creditorInitiatedTransfersEnabled";
    LedgerConfigKey["DomainResolutionFromHandleEnabled"] = "domain.resolutionFromHandleEnabled";
    LedgerConfigKey["AnchorWalletRequired"] = "anchor.walletRequired";
    LedgerConfigKey["PolicyScheduleProcessorEnabled"] = "policy.scheduleProcessorEnabled";
    LedgerConfigKey["TrapsEnabled"] = "traps.enabled";
    LedgerConfigKey["RequestJournalingEnabled"] = "request.journalingEnabled";
    LedgerConfigKey["ForwardAnchorCreate"] = "forward.anchor.create";
    LedgerConfigKey["ForwardAnchorUpdate"] = "forward.anchor.update";
    LedgerConfigKey["ForwardAnchorDrop"] = "forward.anchor.drop";
    LedgerConfigKey["ForwardAnchorGet"] = "forward.anchor.get";
    LedgerConfigKey["ForwardAnchorSign"] = "forward.anchor.sign";
    LedgerConfigKey["ForwardAnchorFindAll"] = "forward.anchor.findAll";
    LedgerConfigKey["AggregatedLimitsEnabled"] = "limits.aggregated.enabled";
    LedgerConfigKey["AuthenticationTwoFactorEnabled"] = "authentication.twoFactor.enabled";
})(LedgerConfigKey || (exports.LedgerConfigKey = LedgerConfigKey = {}));
