// @minka/ledger-sdk/src/lib/effect/effect-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EffectClient = exports.EffectsResponse = exports.EffectResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const access_client_1 = require("../common/clients/access-client");
const base_change_client_1 = require("../common/clients/base-change-client");
const base_client_1 = require("../common/clients/base-client");
const create_record_builder_1 = require("../common/services/create-record-builder");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const list_response_1 = require("../common/types/list-response");
const EFFECTS_PATH = '/effects';
/**
 * Single effect API response.
 */
class EffectResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    effect;
    constructor(response) {
        super(response);
        this.effect = response?.data?.data;
    }
}
exports.EffectResponse = EffectResponse;
/**
 * Effects list API response.
 */
class EffectsResponse extends list_response_1.BaseListResponse {
    effects;
    constructor(response) {
        super(response);
        this.effects = response?.data?.data?.map((record) => record.data);
    }
}
exports.EffectsResponse = EffectsResponse;
/**
 * Allows querying and managing effects.
 *
 * @see {LedgerEffect}
 */
class EffectClient extends base_client_1.BaseClient {
    /**
     * Allows querying changes.
     *
     */
    change;
    /**
     * Allows checking access.
     *
     */
    access;
    constructor(options) {
        super(options);
        this.change = new base_change_client_1.BaseChangeClient(options, EFFECTS_PATH);
        this.access = new access_client_1.AccessClient(options, EFFECTS_PATH);
        this.links = [this.change, this.access];
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     *
     * @see {LedgerEffect}
     * @see {CreateRecordBuilder}
     * @param record initial record payload (optional)
     * @returns a builder instance
     */
    init(record) {
        const builder = new create_record_builder_1.CreateRecordBuilder(async (record, reqOptions) => {
            return this.createRecord(EFFECTS_PATH, EffectResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You can pass the existing record payload to
     * this method to use as initial content.
     *
     * @see {LedgerEffect}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(`${EFFECTS_PATH}/${encodeURIComponent(record.data.handle)}`, EffectResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${EFFECTS_PATH}/${encodeURIComponent(record.data.handle)}/proofs`, EffectResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Gets a ledger effect by identifier (luid or handle).
     *
     * @see {LedgerEffect}
     * @throws {Error} if the effect doesn't exist or the
     *   current user doesn't have access to this effect
     * @param id a unique effect handle
     * @returns a single effect response with the requested effect
     */
    async read(id, options) {
        return this.getRecord(`${EFFECTS_PATH}/${encodeURIComponent(id)}`, EffectResponse, options);
    }
    /**
     * Queries effects from a ledger.
     *
     * @see {LedgerEffect}
     * @param params query parameters
     * @returns effects list response with all found effects
     */
    async list(params) {
        return this.getRecordsList(EFFECTS_PATH, EffectsResponse, params);
    }
}
exports.EffectClient = EffectClient;
